### ISSUE TITLE
Memory leaks

### ISSUE THREAD

**Issue by Medo42** [wareya/faucmix#3]
The project uses very... oldschool memory management, and perhaps as a consequence it looks to me like there are memory leaks. For example, https://github.com/wareya/faucmix/blob/master/src/wavstream.cpp#L16 should really deallocate those buffers that the stream may contain.
IMO the fix should be to use more automatic memory management (e.g. unique_ptr and shared_ptr).

**Comment by wareya** [comment]
Is that the right link?

Now, the downmix_buffer and resample_buffer need to be freed when the wavstream dies, that's true. They should only leak when destroying emitters, though.

In my opinion, faucmix doesn't contain enough manual memory management to burden it with smart pointers. They present a significant barrier of entry to any non-C++ programmers that might ever want to customize faucmix to their own specific needs, or port it to another language, if that ever happens. It should be enough to manually hunt down memory leaks as they happen.

**Comment by Medo42** [comment]
I disagree. Unless there are really specific needs, raw pointers / non-raii
resources should be avoided in modern c++ code. Smart pointers are not hard
to learn and they make it significantly harder to fuck up. If people have
to learn them to mess with faucmix then I consider that a good thing, just
like with that loop that you left uncommented in purpose.

>

**Comment by wareya** [comment]
I want to replace the raw malloc/free usage for buffers of samples by using RAII objects stored in the wavstream. I have no problem with removing malloc/free from logic. Smart pointers only pose a problem in that the way C++ implements them is unlike anything else, even Rust, but that's still a significant problem to me.

As far as I can tell, fauxmix would need to use array smart pointers, which is proscribed: http://www.codingstandard.com/rule/17-3-4-do-not-create-smart-pointers-of-array-type/ It recommends using std::array or std::vector instead; however, my experience trying to use the STL for arrays of audio samples have always resulted in performance problems, even when using references everywhere and compiling in C++17 mode.

**Comment by Medo42** [comment]
That really sounds weird, because there should only be one allocation when the vector is built with the right initial capacity. How bad of a performance problem are we talking? It seems like it would have to be crazy bad to be noticable in this use...

**Comment by wareya** [comment]
It was with an FFT implementation, where the vector was getting read from and written to (a total of n*lb(n) times) in all places by a recursive function. I replaced it with a memory heap which was much faster, then later with std::array which was also slow.

Mingw-w64 had a buggy implementation of std::vector for a while, but I think that it had been fixed by the time I had the FFT implementation, and I think it was only a problem with resizing, not fixed size.

**Comment by wareya** [comment]
Moved wavstream's buffer resizing logic to a new wavbuffer class, which owns the buffer, reallocates it as necessary, deletes it on destruction, and cannot be copied. It's like a special kind of heap class, but can't be copied (unlike STL heap classes), doesn't have iterators, and allows direct access to the underlying array for when it makes sense (unlike a unique pointer).

I know this is basically like using a unique pointer and converting it to a shared pointer, but the buffers are still owned entirely by the wavbuffer. Nothing should hold on to wavstream's output indefinitely, which a shared pointer would allow anyway, and in fact holding onto it indefinitely would be a memory leak itself. Copying into separate scratch buffers that the mixer owns for each emitter would "fix" this, but it would be a waste of performance.

https://github.com/wareya/faucmix/commit/9fb93d271c1d2159feb05611d7b6d1c9b0a8a172
